File: frontend/src/lib/validation.ts
import { FormField } from './types';

// Mirroring the backend verhoeff_checksum
export function verhoeffChecksum(aadhaar: string): boolean {
  if (aadhaar.length !== 12 || !/^\d+$/.test(aadhaar)) return false;

  const d = [
    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
    [1, 2, 3, 4, 0, 6, 7, 8, 9, 5],
    [2, 3, 4, 0, 1, 7, 8, 9, 5, 6],
    [3, 4, 0, 1, 2, 8, 9, 5, 6, 7],
    [4, 0, 1, 2, 3, 9, 5, 6, 7, 8],
    [5, 9, 8, 7, 6, 0, 4, 3, 2, 1],
    [6, 5, 9, 8, 7, 1, 0, 4, 3, 2],
    [7, 6, 5, 9, 8, 2, 1, 0, 4, 3],
    [8, 7, 6, 5, 9, 3, 2, 1, 0, 4],
    [9, 8, 7, 6, 5, 4, 3, 2, 1, 0]
  ];
  const p = [
    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
    [1, 5, 7, 6, 2, 8, 3, 0, 9, 4],
    [5, 8, 0, 3, 7, 9, 6, 1, 4, 2],
    [8, 9, 1, 6, 0, 4, 3, 5, 2, 7],
    [9, 4, 5, 3, 1, 2, 6, 8, 7, 0],
    [4, 2, 8, 6, 5, 7, 3, 9, 0, 1],
    [2, 7, 9, 3, 8, 0, 6, 4, 1, 5],
    [7, 0, 4, 6, 9, 1, 3, 2, 5, 8]
  ];

  let c = 0;
  const reversedAadhaar = aadhaar.split('').reverse();

  for (let i = 0; i < reversedAadhaar.length; i++) {
    c = d[c][p[i % 8][parseInt(reversedAadhaar[i], 10)]];
  }
  return c === 0;
}

export function validateField(field: FormField, value: any, lang: string): string | null {
  const isReq = field.is_required;
  const valStr = value !== undefined && value !== null ? String(value).trim() : '';

  if (!valStr) {
    if (isReq) return lang === 'gu' ? 'આ ફિલ્ડ ફરજિયાત છે.' : lang === 'hi' ? 'यह फ़ील्ड अनिवार्य है।' : 'This field is required.';
    return null;
  }

  // HTML / Control check (simple version)
  if (/<[^>]*>|[<>]/.test(valStr)) {
    return lang === 'gu' ? 'HTML ટેગ્સ માન્ય નથી.' : lang === 'hi' ? 'HTML टैग की अनुमति नहीं है।' : 'HTML/Script tags are not allowed.';
  }

  // Length caps
  if (field.field_type === 'text' && valStr.length > 200) {
    return lang === 'gu' ? '200 અક્ષરોથી વધુ માન્ય નથી.' : lang === 'hi' ? '200 से अधिक वर्णों की अनुमति नहीं है।' : 'Cannot exceed 200 characters.';
  }
  if (field.field_type === 'textarea' && valStr.length > 1000) {
    return lang === 'gu' ? '1000 અક્ષરોથી વધુ માન્ય નથી.' : lang === 'hi' ? '1000 से अधिक वर्णों की अनुमति नहीं है।' : 'Cannot exceed 1000 characters.';
  }

  // Regex
  if (field.validation_regex) {
    const re = new RegExp(field.validation_regex);
    if (!re.test(valStr)) {
      return lang === 'gu' ? 'માહિતી યોગ્ય ફોર્મેટમાં નથી.' : lang === 'hi' ? 'डेटा सही प्रारूप में नहीं है।' : 'Invalid format.';
    }
  }

  // Specific validation metadata
  const meta = field.validation || {};

  if (field.field_type === 'number' || meta.is_decimal || meta.is_integer) {
    const num = Number(valStr);
    if (isNaN(num)) {
      return lang === 'gu' ? 'ફક્ત સંખ્યાઓ માન્ય છે.' : lang === 'hi' ? 'केवल संख्याएँ मान्य हैं।' : 'Only numbers allowed.';
    }
    if (meta.is_integer && !Number.isInteger(num)) {
      return lang === 'gu' ? 'ફક્ત પૂર્ણાંક સંખ્યા માન્ય છે.' : lang === 'hi' ? 'केवल पूर्णांक संख्या मान्य है।' : 'Only integer allowed.';
    }
    if (meta.min !== undefined && num < meta.min) {
      return lang === 'gu' ? `કિંમત ${meta.min} થી ઓછી ન હોવી જોઈએ.` : lang === 'hi' ? `मूल्य ${meta.min} से कम नहीं होना चाहिए।` : `Minimum value is ${meta.min}.`;
    }
    if (meta.max !== undefined && num > meta.max) {
      return lang === 'gu' ? `કિંમત ${meta.max} થી વધુ ન હોવી જોઈએ.` : lang === 'hi' ? `मूल्य ${meta.max} से अधिक नहीं होना चाहिए।` : `Maximum value is ${meta.max}.`;
    }
  }

  if (meta.class === 'aadhaar') {
    if (valStr.length !== 12 || !/^\d+$/.test(valStr) || !verhoeffChecksum(valStr) || new Set(valStr).size === 1) {
      return lang === 'gu' ? 'અમાન્ય આધાર નંબર.' : lang === 'hi' ? 'अमान्य आधार नंबर।' : 'Invalid Aadhaar number.';
    }
  }

  if (field.field_type === 'date' || meta.class === 'dob') {
    const dt = new Date(valStr);
    if (isNaN(dt.getTime())) {
      return lang === 'gu' ? 'અમાન્ય તારીખ ફોર્મેટ.' : lang === 'hi' ? 'अमान्य तिथि प्रारूप।' : 'Invalid date format.';
    }
    if (meta.max_date === 'today' && dt > new Date()) {
      return lang === 'gu' ? 'ભવિષ્યની તારીખ માન્ય નથી.' : lang === 'hi' ? 'भविष्य की तारीख मान्य नहीं है।' : 'Future date is not allowed.';
    }
    if (meta.min_year && dt.getFullYear() < meta.min_year) {
      return lang === 'gu' ? 'અમાન્ય જન્મ તારીખ.' : lang === 'hi' ? 'अमान्य जन्म तिथि।' : 'Invalid Date of Birth.';
    }
  }

  return null;
}

File: backend/formseva_app/core/validation.py
import re
import unicodedata
from datetime import datetime
from typing import Dict, Any, List

def verhoeff_checksum(aadhaar: str) -> bool:
    """Verhoeff algorithm for Aadhaar validation."""
    if len(aadhaar) != 12 or not aadhaar.isdigit():
        return False
        
    d = (
        (0, 1, 2, 3, 4, 5, 6, 7, 8, 9),
        (1, 2, 3, 4, 0, 6, 7, 8, 9, 5),
        (2, 3, 4, 0, 1, 7, 8, 9, 5, 6),
        (3, 4, 0, 1, 2, 8, 9, 5, 6, 7),
        (4, 0, 1, 2, 3, 9, 5, 6, 7, 8),
        (5, 9, 8, 7, 6, 0, 4, 3, 2, 1),
        (6, 5, 9, 8, 7, 1, 0, 4, 3, 2),
        (7, 6, 5, 9, 8, 2, 1, 0, 4, 3),
        (8, 7, 6, 5, 9, 3, 2, 1, 0, 4),
        (9, 8, 7, 6, 5, 4, 3, 2, 1, 0)
    )
    p = (
        (0, 1, 2, 3, 4, 5, 6, 7, 8, 9),
        (1, 5, 7, 6, 2, 8, 3, 0, 9, 4),
        (5, 8, 0, 3, 7, 9, 6, 1, 4, 2),
        (8, 9, 1, 6, 0, 4, 3, 5, 2, 7),
        (9, 4, 5, 3, 1, 2, 6, 8, 7, 0),
        (4, 2, 8, 6, 5, 7, 3, 9, 0, 1),
        (2, 7, 9, 3, 8, 0, 6, 4, 1, 5),
        (7, 0, 4, 6, 9, 1, 3, 2, 5, 8)
    )
    inv = (0, 4, 3, 2, 1, 5, 6, 7, 8, 9)
    
    c = 0
    reversed_aadhaar = aadhaar[::-1]
    
    for i, char in enumerate(reversed_aadhaar):
        c = d[c][p[i % 8][int(char)]]
        
    return c == 0

def has_html(text: str) -> bool:
    """Check if string contains HTML tags or angle brackets."""
    return bool(re.search(r'<[^>]*>|[<>]', text))

def has_control_chars(text: str) -> bool:
    """Check if string contains control characters."""
    return any(unicodedata.category(c) == 'Cc' for c in text if c not in ('\n', '\r', '\t'))

def normalize_text(text: str) -> str:
    """Trim and Unicode normalize (NFC)."""
    return unicodedata.normalize('NFC', str(text).strip())

class ValidationError(Exception):
    def __init__(self, field_key: str, message_gu: str, message_hi: str, message_en: str):
        self.field_key = field_key
        self.message_gu = message_gu
        self.message_hi = message_hi
        self.message_en = message_en
        super().__init__(self.message_en)

def validate_form_fields(
    payload_fields: Dict[str, Any], 
    form_def: Dict[str, Any], 
    form_fields: List[Dict[str, Any]]
) -> Dict[str, Any]:
    """
    Validates payload against form field definitions.
    Returns cleaned payload. Raises ValidationError on first failure, 
    or can be modified to return all errors.
    """
    cleaned_data = {}
    
    # 1. Whitelist Keys
    allowed_keys = {f['field_key']: f for f in form_fields}
    
    for key, raw_val in payload_fields.items():
        if key not in allowed_keys:
            raise ValidationError(
                key,
                f"ફિલ્ડ '{key}' આ ફોર્મનો ભાગ નથી.",
                f"फ़ील्ड '{key}' इस फ़ॉर्म का हिस्सा नहीं है।",
                f"Field '{key}' is not part of this form."
            )
            
    # 2. Check Required Fields & Validate Values
    for field_def in form_fields:
        key = field_def['field_key']
        is_req = field_def.get('is_required', False)
        
        if key not in payload_fields or payload_fields[key] is None or str(payload_fields[key]).strip() == "":
            if is_req:
                raise ValidationError(key, "આ ફિલ્ડ ફરજિયાત છે.", "यह फ़ील्ड अनिवार्य है।", "This field is required.")
            continue
            
        val = str(payload_fields[key])
        
        # UNIVERSAL RULES
        val = normalize_text(val)
        
        if has_control_chars(val):
            raise ValidationError(key, "અમાન્ય અક્ષરો મળ્યા.", "अमान्य वर्ण मिले।", "Control characters are not allowed.")
            
        if has_html(val):
            raise ValidationError(key, "HTML ટેગ્સ માન્ય નથી.", "HTML टैग की अनुमति नहीं है।", "HTML/Script tags are not allowed.")
            
        field_type = field_def.get('field_type', 'text')
        
        # MAX LENGTH CAPS BY TYPE
        if field_type == 'text' and len(val) > 200:
             raise ValidationError(key, "200 અક્ષરોથી વધુ માન્ય નથી.", "200 से अधिक वर्णों की अनुमति नहीं है।", "Cannot exceed 200 characters.")
        if field_type == 'textarea' and len(val) > 1000:
             raise ValidationError(key, "1000 અક્ષરોથી વધુ માન્ય નથી.", "1000 से अधिक वर्णों की अनुमति नहीं है।", "Cannot exceed 1000 characters.")
             
        # OPTIONS / SELECT VALIDATION
        if field_type in ['select', 'radio']:
            options = field_def.get('options_json', [])
            valid_values = [opt.get('value') for opt in options] if isinstance(options, list) else []
            if valid_values and val not in valid_values:
                 raise ValidationError(key, "પસંદ કરેલ વિકલ્પ અમાન્ય છે.", "चयनित विकल्प अमान्य है।", "Selected option is invalid.")
                 
        # REGEX VALIDATION
        val_regex = field_def.get('validation_regex')
        if val_regex:
            if not re.match(val_regex, val):
                raise ValidationError(key, "માહિતી યોગ્ય ફોર્મેટમાં નથી.", "डेटा सही प्रारूप में नहीं है।", "Invalid format.")
                
        # SPECIFIC VALIDATION METADATA
        val_meta = field_def.get('validation') or {}
        
        if field_type == 'number' or val_meta.get('is_decimal') or val_meta.get('is_integer'):
            try:
                num_val = float(val)
                if val_meta.get('is_integer') and not float(val).is_integer():
                    raise ValidationError(key, "ફક્ત પૂર્ણાંક સંખ્યા માન્ય છે.", "केवल पूर्णांक संख्या मान्य है।", "Only integer allowed.")
                    
                min_val = val_meta.get('min')
                max_val = val_meta.get('max')
                
                if min_val is not None and num_val < min_val:
                    raise ValidationError(key, f"કિંમત {min_val} થી ઓછી ન હોવી જોઈએ.", f"मूल्य {min_val} से कम नहीं होना चाहिए।", f"Minimum value is {min_val}.")
                if max_val is not None and num_val > max_val:
                    raise ValidationError(key, f"કિંમત {max_val} થી વધુ ન હોવી જોઈએ.", f"मूल्य {max_val} से अधिक नहीं होना चाहिए।", f"Maximum value is {max_val}.")
            except ValueError:
                raise ValidationError(key, "ફક્ત સંખ્યાઓ માન્ય છે.", "केवल संख्याएँ मान्य हैं।", "Only numbers allowed.")
                
        # AADHAAR
        if val_meta.get('class') == 'aadhaar':
            if len(val) != 12 or not val.isdigit() or not verhoeff_checksum(val) or len(set(val)) == 1:
                raise ValidationError(key, "અમાન્ય આધાર નંબર.", "अमान्य आधार नंबर।", "Invalid Aadhaar number.")
                
        # DOB / DATE
        if field_type == 'date' or val_meta.get('class') == 'dob':
            try:
                dt = datetime.strptime(val, "%Y-%m-%d")
                if val_meta.get('max_date') == 'today' and dt.date() > datetime.today().date():
                    raise ValidationError(key, "ભવિષ્યની તારીખ માન્ય નથી.", "भविष्य की तारीख मान्य नहीं है।", "Future date is not allowed.")
                if val_meta.get('min_year') and dt.year < val_meta.get('min_year'):
                    raise ValidationError(key, "અમાન્ય જન્મ તારીખ.", "अमान्य जन्म तिथि।", "Invalid Date of Birth.")
            except ValueError:
                # Fallback format checking
                try:
                    dt = datetime.strptime(val, "%d/%m/%Y")
                    val = dt.strftime("%Y-%m-%d") # Normalize format
                except ValueError:
                    raise ValidationError(key, "અમાન્ય તારીખ ફોર્મેટ.", "अमान्य तिथि प्रारूप।", "Invalid date format.")
                    
        # CROSS FIELD (Example EWS)
        if val_meta.get('max_ews'):
            if form_def.get('slug') == 'ews_certificate':
                try:
                    num_val = float(val)
                    if num_val > val_meta.get('max_ews'):
                        raise ValidationError(key, "EWS માટે આવક મર્યાદા કરતા વધારે છે.", "ईडब्ल्यूएस के लिए आय सीमा से अधिक है।", "Income exceeds EWS limits.")
                except ValueError:
                    pass

        cleaned_data[key] = val
        
    return cleaned_data
